cpp/src/io/parquet/writer_impl.cu

```diff
@@ -381,133 +399,151 @@ struct leaf_schema_fn {
-      col_schema.logical_type.TIMESTAMP.unit.isset.NANOS = true;
+      col_schema.logical_type                 = LogicalType{LogicalType::TIMESTAMP};
+      col_schema.logical_type->timestamp_type = TimestampType{false, TimeUnit::NANOS};
     }
   }
 
   //  unsupported outside cudf for parquet 1.0.
   template <typename T>
   std::enable_if_t<std::is_same_v<T, cudf::duration_D>, void> operator()()
   {
-    col_schema.type                                = Type::INT32;
-    col_schema.converted_type                      = ConvertedType::TIME_MILLIS;
-    col_schema.stats_dtype                         = statistics_dtype::dtype_int32;
-    col_schema.ts_scale                            = 24 * 60 * 60 * 1000;
-    col_schema.logical_type.isset.TIME             = true;
-    col_schema.logical_type.TIME.unit.isset.MILLIS = true;
+    col_schema.type                    = Type::INT32;
+    col_schema.converted_type          = ConvertedType::TIME_MILLIS;
+    col_schema.stats_dtype             = statistics_dtype::dtype_int32;
+    col_schema.ts_scale                = 24 * 60 * 60 * 1000;
+    col_schema.logical_type            = LogicalType{LogicalType::TIME};
+    col_schema.logical_type->time_type = TimeType{false, TimeUnit::MILLIS};
   }
 
   template <typename T>
   std::enable_if_t<std::is_same_v<T, cudf::duration_s>, void> operator()()
   {
-    col_schema.type                                = Type::INT32;
-    col_schema.converted_type                      = ConvertedType::TIME_MILLIS;
-    col_schema.stats_dtype                         = statistics_dtype::dtype_int32;
-    col_schema.ts_scale                            = 1000;
-    col_schema.logical_type.isset.TIME             = true;
-    col_schema.logical_type.TIME.unit.isset.MILLIS = true;
+    col_schema.type                    = Type::INT32;
+    col_schema.converted_type          = ConvertedType::TIME_MILLIS;
+    col_schema.stats_dtype             = statistics_dtype::dtype_int32;
+    col_schema.ts_scale                = 1000;
+    col_schema.logical_type            = LogicalType{LogicalType::TIME};
+    col_schema.logical_type->time_type = TimeType{false, TimeUnit::MILLIS};
   }
 
   template <typename T>
   std::enable_if_t<std::is_same_v<T, cudf::duration_ms>, void> operator()()
   {
-    col_schema.type                                = Type::INT32;
-    col_schema.converted_type                      = ConvertedType::TIME_MILLIS;
-    col_schema.stats_dtype                         = statistics_dtype::dtype_int32;
-    col_schema.logical_type.isset.TIME             = true;
-    col_schema.logical_type.TIME.unit.isset.MILLIS = true;
+    col_schema.type                    = Type::INT32;
+    col_schema.converted_type          = ConvertedType::TIME_MILLIS;
+    col_schema.stats_dtype             = statistics_dtype::dtype_int32;
+    col_schema.logical_type            = LogicalType{LogicalType::TIME};
+    col_schema.logical_type->time_type = TimeType{false, TimeUnit::MILLIS};
```

Comment:
We could have constructors to avoid multi-step initialization
LogicalType(TimeType) etc.

Reply:
yeah, I guess that would get rid of some of the anxiety with type being set, but the struct being left `nullopt`.